File: exercises/C#/04 CruiseControl/04 CruiseControl/Hardware.cs
using System;

namespace _04_CruiseControl
{
    class Hardware
    {
        private static Hardware _instance = null;

        private Hardware()
        {
        }

        public static Hardware GetInstance()
        {
            if (_instance == null)
            {
                _instance = new Hardware();
            }

            return _instance;
        }

        public void Accelerate(bool accelerate)
        {
            // Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
            throw new Exception("No connection to real car");
        }

        public int GetCurrentSpeed()
        {
            // Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
            throw new Exception("No connection to real car");
        }

        public bool IsBrakePressed()
        {
            // Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
            throw new Exception("No connection to real car");
        }

        public bool IsClutchPressed()
        {
            // Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
            throw new Exception("No connection to real car");
        }

        public bool IsCruiseControlOn()
        {
            // Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
            throw new Exception("No connection to real car");
        }

        public void DisableCruiseControl()
        {
            // Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
            throw new Exception("No connection to real car");
        }

        public int GetCruiseControlSpeed()
        {
            // Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
            throw new Exception("No connection to real car");
        }
    }
}

File: exercises/04 CruiseControl/C++/04 CruiseControl/Hardware.cpp
#include "Hardware.h"
#include <stdexcept>

using namespace std;

Hardware* Hardware::_instance = nullptr;

Hardware& Hardware::getInstance()
{
	if (_instance == nullptr)
	{
		_instance = new Hardware();
	}

	return *_instance;
}

void Hardware::accelerate(bool accelerate)
{
	// Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
	throw logic_error("No connection to real car");
}

int Hardware::getCurrentSpeed() const
{
	// Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
	throw logic_error("No connection to real car");
}

bool Hardware::isBreakPressed() const
{
	// Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
	throw logic_error("No connection to real car");
}

bool Hardware::isClutchPressed() const
{
	// Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
	throw logic_error("No connection to real car");
}

bool Hardware::isCruiseControlOn() const
{
	// Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
	throw logic_error("No connection to real car");
}

void Hardware::disableCruiseControl()
{
	// Don't change the body of this method. Consider this method containing logic for actually controlling the hardware.
	throw logic_error("No connection to real car");
}

int Hardware::getCruiseControlSpeed() const
{
	throw logic_error("No connection to real car");
}

Hardware::Hardware()
{
}


Hardware::~Hardware()
{
}
